src/vm/memory.zig

```diff
@@ -104,71 +142,232 @@ pub const Region = struct {
+        const reg = self.find_region(vm_addr) orelse return err;
+        return reg.translate(state, vm_addr, len) orelse return err;
     }
 };
 
+// [agave] https://github.com/anza-xyz/sbpf/blob/a8247dd30714ef286d26179771724b91b199151b/src/memory_region.rs#L870
+fn accessViolation(
+    vm_addr: u64,
+    version: sbpf.Version,
+    config: *const exe.Config,
+) AccessError {
+    const stack_frame_idx = std.math.divExact(
+        u64,
+        vm_addr -| STACK_START,
+        config.stack_frame_size,
+    ) catch 0;
+
+    return if (!version.enableDynamicStackFrames() and stack_frame_idx < config.max_call_depth)
+        error.StackAccessViolation
+    else
+        error.AccessViolation;
+}
+
+// [agave] https://github.com/anza-xyz/sbpf/blob/a8247dd30714ef286d26179771724b91b199151b/src/memory_region.rs#L183
+/// Memory mapping based on eytzinger search.
+const UnalignedMemoryMap = struct {
+    /// Mapped memory regions
+    regions: []Region,
+    /// Copy of the regions vm_addr fields to improve cache density
+    region_addresses: []u64,
+    // Executable sbpf_version
+    version: sbpf.Version,
+    // VM configuration
+    config: *const exe.Config,
+
+    // CoW callback
+    //cow_cb: ?MemoryCowCallback,
+    // Cache of the last `MappingCache::SIZE` vm_addr => region_index lookups
+    //cache: MappingCache,
+
+    // [agave] https://github.com/anza-xyz/sbpf/blob/a8247dd30714ef286d26179771724b91b199151b/src/memory_region.rs#L241
+    fn init(
+        allocator: std.mem.Allocator,
+        _regions: []const Region,
+        version: sbpf.Version,
+        config: *const exe.Config,
+    ) (error{OutOfMemory} || InitError)!UnalignedMemoryMap {
+        const sorted_regions = try allocator.dupe(Region, _regions);
+        defer allocator.free(sorted_regions);
+        std.mem.sort(Region, sorted_regions, {}, Region.lessThanFn);
+
+        if (Region.regionsOverlap(sorted_regions)) return error.InvalidMemoryRegion;
+
+        const region_addresses = try allocator.alloc(u64, sorted_regions.len);
+        errdefer allocator.free(region_addresses);
+
+        const regions = try allocator.alloc(Region, _regions.len);
+        errdefer allocator.free(regions);
+
+        var self: UnalignedMemoryMap = .{
+            .regions = regions,
+            .region_addresses = region_addresses,
+            // .cache
+            .version = version,
+            .config = config,
+            // .cow_cb
+        };
+
+        _ = self.constructEytzingerOrder(sorted_regions, 0, 0);
+
+        return self;
+    }
+
+    fn deinit(self: *const UnalignedMemoryMap, allocator: std.mem.Allocator) void {
+        allocator.free(self.regions);
+        allocator.free(self.region_addresses);
+    }
+
+    // [agave] https://github.com/anza-xyz/sbpf/blob/a8247dd30714ef286d26179771724b91b199151b/src/memory_region.rs#L218
+    fn constructEytzingerOrder(
+        self: UnalignedMemoryMap,
+        ascending_regions: []Region,
+        _in_index: usize,
```

Comment:
is there no better var names for the vars with the _ prefix? code can become hard to read with too many

Reply:
renamed using in_index and new_in_index https://github.com/Syndica/sig/pull/639/commits/e2a0e3c58e0f908854d59b9212eca28bc5f1bd0e#diff-23e4e8b8c3b0ebce6d44fff7c7ff93546043dcee6a94c6c2b54d5482d0457468R305-R321